### ISSUE TITLE
heredoc backwards compatibility with HCL

### ISSUE THREAD

**Issue by seanvaidya** [hashicorp/nomad#11491]
### Nomad version
Incrementally upgrading from 0.8.7 to 1.1.6; issues started with 0.12.12

### Operating system and Environment details
AWS, Amazon linux 2
Local machine is Ubuntu

### Issue
On 0.8.7 we are able to get way with the identifier (e.g. EOF) not being on a separate line. The use case for us was needing to prevent a newline character from beind added. In my 0.12.12 upgrade, HCL2 was being used to render and it requires that the EOF be on its own line. 

This is fine because I can just use trimspace to remove the newline character. The problem, however, is that I am upgrading in one environment and at the same time I need to be able to operate in another environment which is still running on 0.8.7 with minimal config drift between the two, though my janky workaround to this issue does involve if statements which selects the config based on the nomad version. 

The solution here *should* be to run in HCL 1 mode (-hcl1) to keep backwards compatibility but that doesn't seem to work. So now I'm stuck with a newline character that I can't trim.

### Reproduction steps
Create a heredoc data and place the EOF identified at the end of the line instead of below in a seperate line. 
Works on 0.8.7, does not work on 0.12.12 with -hcl1

#### Expected Result
-hcl1 should provide full backwards compatibility.

#### Actual Result
Newline that cannot be trimmed. Compatibility borked.

**Comment by tgross** [comment]
Hi @seanvaidya! Upgrading all the way from 0.8.x to 1.1.x is quite an operation! Just to make sure there's no ambiguity translating from English to HCL 😀 , can you provide an example jobspec with heredoc data that exposes this problem? 

> The solution here _should_ be to run in HCL 1 mode (-hcl1) to keep backwards compatibility but that doesn't seem to work. So now I'm stuck with a newline character that I can't trim.

The `-hcl1` mode tells the parser to use the HCL1 code path. But we didn't "freeze" that code path, as we needed to be able to support new features in HCL1 for some time (at some point we'll deprecate HCL1 entirely of course). So if the HCL1 code path changed then that could potentially result in subtle changes. I do want to point out that if we were to update this, the fix won't land in the 0.12.x series as that series is no longer getting backports.

**Comment by seanvaidya** [comment]
I'll do you one better, here's the (janky)  work-around:

`      template {
        [[if eq .NEW_NOMAD_VERSION "true" ]]
        data = trimspace(<<EOF
{{ with secret "secrets/test/creds" }}{{ .Data.testa_key | base64Decode }}{{ end }}
EOF
)
        [[else]]
        data = <<EOF
{{ with secret "secrets/test/creds" }}{{ .Data.testa_key | base64Decode }}{{ end }}EOF
        [[end]]
      destination = "[[.TEST_PATH]]"
      }
`

**Comment by tgross** [comment]
Hi @seanvaidya! So unfortunately I wasn't able to make a replication out of that, which is why I was really asking for a specific piece of text that shows the problem with HCL2 vs HCL1 parsing. It also looks like you've got Levant templating (which doesn't accept HCL2 syntax) and Vault secrets decoding in the mix, which makes the results a little harder to pick apart.

I first tried the following minimal job:

```hcl
job "example" {
  datacenters = ["dc1"]
  group "g1" {
    task "t1" {
      driver = "docker"
      config {
        image   = "busybox:1"
        command = "/bin/sh"
        args    = ["-c", "sleep 300"]
      }

      template {
        data = <<EOF
text version 1 with no trailing newlineEOF
        destination = "local/text1.txt"
      }

      template {
        data = <<EOF
text version 2 with trailing newline
EOF
        destination = "local/text2.txt"
      }
    }
  }
}
```

But as it turns out the EOF without a newline isn't accepted as an EOF marker at all by Nomad on either parser! For both HCL1 and HCL2 I get the following:

```
$ nomad job run -hcl1 ./example.nomad
...
$ nomad alloc fs 5c5 t1/local
Mode        Size   Modified Time         Name
-rw-r--r--  163 B  2021-11-12T15:23:19Z  text2.txt
$ nomad alloc fs 5c5 t1/local/text2.txt
text version 1 with no trailing newlineEOF
        destination = "local/text1.txt"
      }

      template {
        data = <<EOF
text version 2 with trailing newline
```

With Nomad 0.12.11, which was the version before where you reported you saw a problem, I got the exact same bug:

```
$ ./nomad_0.12.11 job run ./example.nomad
...
$ ./nomad_0.12.11 alloc fs b5c t1/local/text2.txt
text version 1 with no trailing newlineEOF
        destination = "local/text1.txt"
      }

      template {
        data = <<EOF
text version 2 with trailing newline
```

If I remove the broken template that's missing a newline before the EOF, I get the following template as expected, with a trailing newline:

```
$ nomad alloc fs aa t1/local/text2.txt
text version 2 with trailing newline

```

And I can use the `trimspace` function in HCL2 as you've done, but of course that function doesn't work in `-hcl1` mode:

```hcl
      template {
        data = trimspace(<<EOF
text version 2 with trailing newline
EOF
          )
        destination = "local/text2.txt"
      }
```

And that results in the following as expected (no trailing newline):

```
$ nomad alloc fs 894 t1/local/text2.txt
text version 2 with trailing newline%

```

I also tried with the following Levant template:

```hcl
      [[ if true ]]

      template {
        data = <<EOF
text version 1 with no trailing newlineEOF
        destination = "local/text1.txt"
      }

      template {
        data = <<EOF
text version 2 with trailing newline
EOF
        destination = "local/text2.txt"
      }

      [[ end ]]
```

And that ends up being unable to parse the EOF as expected either:

```
$ nomad alloc fs 2b0 t1/local/text2.txt
text version 1 with no trailing newlineEOF
        destination = "local/text1.txt"
      }

      template {
        data = <<EOF
text version 2 with trailing newline

```

So tl;dr is that unfortunately I don't have a good replication of the problem you're having, as it doesn't look like having a missing newline before the EOF is accepted by any of the versions of Nomad I've tried, including passing through Levant.

While I appreciate you trying to show me the workaround, do you have an example of `template` block that just includes a string literal without Levant templating that demonstrates the issue?

**Comment by tgross** [comment]
Also, maybe a dumb question, but given that you're blitting a secret out to a file anyways, why have a heredoc at all? Could you do:

```hcl
template {
  data = "{{- with secret \"secrets/test/creds\" }}{{ .Data.testa_key | base64decode }}{{ end -}}"
  destination = "secret/creds"
}
```

**Comment by seanvaidya** [comment]
Apologies for the late reply-- 

I started noticing the issue in 0.12.12, however for a successful reproduction you likely need to use the last major version of 0.11 as I'm currently upgrading one major version at time from 0.8 to 1.0 so I never even tried 0.12.11.

As for your second comment, that most likely would work but I think this bug is still relevant because ideally we'd be able to get away with using mostly the same config that worked in an old version with -hcl1.

I think you can close this as there may not be anything to fix. But the info within may be useful for someone in a similar situation.

**Comment by tgross** [comment]
Hi @seanvaidya! Ok, so I wasn't able to reproduce on 0.11.8 either. But given that we're back that far at this point and into unsupported versions of Nomad that aren't receiving backports, there's also not much we could do to fix the behavior at this point without potentially breaking the current behavior for users. I'm going to close this issue as you've suggested, but thanks for reporting this and for being patient while we tried to reproduce.
